### PR TITLE
added test for dont cache empty results

### PR DESCRIPTION
**Pull request by shubham-shay-te**
fixed a warning for `assertEquals` to `assertEqual` for default registry test

### CODE REVIEW

File: tests/test_cache.py [hunk 1]
```diff
@@ -195,6 +200,138 @@ def test_cache_using_string(self):
...
+        self.assertSetEqual(self.get_cache_empty(TestCache.test_set), {1.0, "Hello", (1, 2, 3)})
+        with freeze_time(datetime.utcnow() + timedelta(seconds=4)):
+            self.assertSetEqual(self.get_cache_empty(TestCache.test_set), set())
+
+    def test_cache_empty_false(self):
+        # Dict
+        self.clear_cache_empty_false()
+        TestCache.test_dict = {}
+        self.assertDictEqual(self.get_cache_empty_false(TestCache.test_dict), {})
+        TestCache.test_dict = {'value': '100'}
```
Comment (shay-te) [inline tests/test_cache.py:274]: `{'value': '100'}` is being reused a lot. we can define it in a variable
suggestion:
`test_value = {'value': '100'}`

File: tests/test_cache.py [hunk 2]
```diff
@@ -282,3 +419,19 @@ def get_cache_expire_string_year(self):
...
+    @Cache(key="test_cache_empty", expire=timedelta(seconds=2))
+    def get_cache_empty(self, param):
+        return param
+
+    @Cache(key="test_cache_empty", invalidate=True)
+    def clear_cache_empty(self):
+        pass
+
+    @Cache(key="test_cache_empty_false", cache_empty_result=False, expire=timedelta(seconds=2))
+    def get_cache_empty_false(self, param):
```
Comment (shay-te) [inline tests/test_cache.py:432]: the name `get_cache_empty_false` is not clear. 
suggestion:
`get_dont_cache_empty_results`

File: tests/test_cache.py [hunk 3]
```diff
@@ -282,3 +419,19 @@ def get_cache_expire_string_year(self):
     @Cache(key="test_cache_expire_string_year", invalidate=True)
     def clear_cache_expire_string_year(self):
         pass
+
+    @Cache(key="test_cache_empty", expire=timedelta(seconds=2))
+    def get_cache_empty(self, param):
+        return param
```
Comment (shay-te) [inline tests/test_cache.py:425]: the name `get_cache_empty` is not clear. 
suggestion:
`get_cache_empty_results`

File: tests/test_cache.py [hunk 4]
```diff
@@ -282,3 +419,19 @@ def get_cache_expire_string_year(self):
     @Cache(key="test_cache_expire_string_year", invalidate=True)
     def clear_cache_expire_string_year(self):
         pass
+
+    @Cache(key="test_cache_empty", expire=timedelta(seconds=2))
```
Comment (shay-te) [inline tests/test_cache.py:423]: lets try this for the entire test. for the reason i want this to be a good example of how to use cache. 

all cache `key` will become a constant that will be reused also in the invalidate 

```Python
CACHE_TEST_CACHE_EMPTY = 'test_cache_empty'
```

File: tests/test_cache.py [hunk 5]
```diff
@@ -195,6 +200,138 @@ def test_cache_using_string(self):
...
+            self.assertListEqual(self.get_cache_empty(TestCache.test_list), [])
+
+        # String
+        self.clear_cache_empty()
+        TestCache.test_string = ""
+        self.assertEqual(self.get_cache_empty(TestCache.test_string), "")
+        TestCache.test_string = "Hello World"
+        self.assertEqual(self.get_cache_empty(TestCache.test_string), "")
+        with freeze_time(datetime.utcnow() + timedelta(seconds=2)):
+            self.assertEqual(self.get_cache_empty(TestCache.test_string), "Hello World")
```
Comment (shay-te) [inline tests/test_cache.py:250]: `"Hello world" into `test_value_str`

File: tests/test_cache.py [hunk 6]
```diff
@@ -195,6 +200,138 @@ def test_cache_using_string(self):
...
+        TestCache.test_tuple = ()
+        self.assertTupleEqual(self.get_cache_empty(TestCache.test_tuple), ('value', '100'))
+        with freeze_time(datetime.utcnow() + timedelta(seconds=4)):
+            self.assertTupleEqual(self.get_cache_empty(TestCache.test_tuple), ())
+
+        # List
+        self.clear_cache_empty()
+        TestCache.test_list = []
+        self.assertListEqual(self.get_cache_empty(TestCache.test_list), [])
+        TestCache.test_list = ['value', '100']
```
Comment (shay-te) [inline tests/test_cache.py:234]: reuse variable `test_value_list = ['value', '100']`

File: tests/test_cache.py [hunk 7]
```diff
@@ -195,6 +200,138 @@ def test_cache_using_string(self):
...
+        TestCache.test_dict = {}
+        self.assertDictEqual(self.get_cache_empty(TestCache.test_dict), {'value': '100'})
+        with freeze_time(datetime.utcnow() + timedelta(seconds=4)):
+            self.assertDictEqual(self.get_cache_empty(TestCache.test_dict), {})
+
+        # Tuple
+        self.clear_cache_empty()
+        TestCache.test_tuple = ()
+        self.assertTupleEqual(self.get_cache_empty(TestCache.test_tuple), ())
+        TestCache.test_tuple = ('value', '100')
```
Comment (shay-te) [inline tests/test_cache.py:221]: reuse variable  `test_value_tuple = ('value', '100')`

File: tests/test_cache.py [hunk 8]
```diff
@@ -195,6 +200,138 @@ def test_cache_using_string(self):
         TestCache.test_value = 100
         self.assertEqual(self.get_cache_expire_string_year(), 100)
 
+    def test_cache_empty(self):
+        # Dict
+        self.clear_cache_empty()
+        TestCache.test_dict = {}
+        self.assertDictEqual(self.get_cache_empty(TestCache.test_dict), {})
+        TestCache.test_dict = {'value': '100'}
```
Comment (shay-te) [inline tests/test_cache.py:208]: reuse variable test_value_dict = {'value': '100'}

File: tests/test_cache.py [hunk 9]
```diff
@@ -195,6 +200,138 @@ def test_cache_using_string(self):
...
+        with freeze_time(datetime.utcnow() + timedelta(seconds=2)):
+            self.assertEqual(self.get_cache_empty(TestCache.test_string), "Hello World")
+        TestCache.test_string = ""
+        self.assertEqual(self.get_cache_empty(TestCache.test_string), "Hello World")
+        with freeze_time(datetime.utcnow() + timedelta(seconds=4)):
+            self.assertEqual(self.get_cache_empty(TestCache.test_string), "")
+
+        # Set
+        self.clear_cache_empty()
+        TestCache.test_set = set()
```
Comment (shay-te) [inline tests/test_cache.py:258]: reuse variable test_value_set= {1.0, "Hello", (1, 2, 3)}

File: tests/test_cache.py [hunk 10]
```diff
@@ -195,6 +200,138 @@ def test_cache_using_string(self):
...
+        self.assertDictEqual(self.get_cache_empty_false(TestCache.test_dict), {'value': '100'})
+        TestCache.test_dict = {}
+        self.assertDictEqual(self.get_cache_empty_false(TestCache.test_dict), {'value': '100'})
+        with freeze_time(datetime.utcnow() + timedelta(seconds=2)):
+            self.assertDictEqual(self.get_cache_empty_false(TestCache.test_dict), {})
+        TestCache.test_dict = {'value': '100'}
+        self.assertDictEqual(self.get_cache_empty_false(TestCache.test_dict), {'value': '100'})
+
+        # Tuple
+        self.clear_cache_empty_false()
```
Comment (shay-te) [inline tests/test_cache.py:284]: this is hard to review. 
`test_cache_empty_false` + `test_cache_empty` can be done the same as the test of result to dict. 
just create one function for testing with cache and another function to test without cache. 
and pass diffrent value every time

then just pass a different variable every time

**Review by shay-te — CHANGES_REQUESTED** [review]

File: website/docs/cache.md [hunk 11]
```diff
@@ -61,17 +61,27 @@ cache_client_factory.get() # returns None. Multiple client registered.
...
     `None`: the decorated  `function.__qualname__` is used.     
     
     `some_key{param_1}{param_2}`: will build a key with the `param_1` and `param_2` values.     
     when a parameter is optional and empty `_` is used. 
+* `max_key_length` default `250`, the maximum length of key string to be accepted by decorator.
 * `expire` Period of time when the value is expired.
 * `invalidate` Remove the value from the cache using the key.
-* `cache_client_name` The name to pay to get the correct `CacheHandler`.
+* `handler` name of the handler to be used for handling the cache, by default `Core-Lib`'s default `CacheHandler` is used.
+* `cache_empty_result` type `bool`, default `True`, is responsible for caching the empty values like `{}`, `[]`, `()`, `""` and `set()`.
```
Comment (shay-te) [inline website/docs/cache.md:84]: `cache_empty_result` type `bool`, default `True`. When `True` will cache empty values as `{}`, `[]`, `()`, `""` and `set()`

File: website/docs/cache.md [hunk 12]
```diff
@@ -61,17 +61,27 @@ cache_client_factory.get() # returns None. Multiple client registered.
...
    
     `None`: the decorated  `function.__qualname__` is used.     
     
     `some_key{param_1}{param_2}`: will build a key with the `param_1` and `param_2` values.     
     when a parameter is optional and empty `_` is used. 
+* `max_key_length` default `250`, the maximum length of key string to be accepted by decorator.
 * `expire` Period of time when the value is expired.
 * `invalidate` Remove the value from the cache using the key.
-* `cache_client_name` The name to pay to get the correct `CacheHandler`.
+* `handler` name of the handler to be used for handling the cache, by default `Core-Lib`'s default `CacheHandler` is used.
```
Comment (shay-te) [inline website/docs/cache.md:83]: the parameter name is `handler_name` . 
this is incorrect, please look at the code. `CoreLib.cache_registry.get(self.handler_name)` 
will specify what CacheHandler to user, using the  `CoreLib.cache_registry`
please use grammerly

File: tests/test_cache.py [hunk 13]
```diff
@@ -195,90 +200,164 @@ def test_cache_using_string(self):
...
+            with freeze_time(datetime.utcnow() + timedelta(seconds=2)):
+                self.assertEqual(self.get_cache_empty_results(cache_data), cache_data)
+            self.assertEqual(self.get_cache_empty_results(empty_value), cache_data)
+            with freeze_time(datetime.utcnow() + timedelta(seconds=4)):
+                self.assertEqual(self.get_cache_empty_results(empty_value), empty_value)
+        else:
+            self.clear_cache_empty_false()
+            self.assertEqual(self.get_dont_cache_empty_results(empty_value), empty_value)
+            self.assertEqual(self.get_dont_cache_empty_results(cache_data), cache_data)
+            TestCache.test_dict = {}
```
Comment (shay-te) [inline tests/test_cache.py:217]: is this related tp the test?

File: tests/test_cache.py [hunk 14]
```diff
@@ -195,90 +200,164 @@ def test_cache_using_string(self):
...
+            self.assertEqual(self.get_dont_cache_empty_results(empty_value), cache_data)
```
Comment (shay-te) [inline tests/test_cache.py:218]: how this is different?

**Review by shay-te — CHANGES_REQUESTED** [review]

File: tests/test_cache.py (hunk 13 above)
Comment (shubham-shay-te) [reply tests/test_cache.py:217]: sorry, my mistake, had to remove that while pushing the code

File: tests/test_cache.py (hunk 14 above)
Comment (shubham-shay-te) [reply tests/test_cache.py:218]: just to check, if the actual cache expires or not after 2 seconds so we give it the `empty_value ` and then `assertEqual` it in the next step after 2 seconds.

File: website/docs/cache.md (hunk 12 above)
Comment (shubham-shay-te) [reply website/docs/cache.md:83]: if you look at the `init` function of `Cache Decorator`, we have the parameter name as `handler` and then afterward we are assigning the `handler` to `self.handler_name` which we use internally in the class

Comment (shay-te) [reply website/docs/cache.md:83]: OK, you are correct. 
the reason is that `handler` can be understood as a class that handle something 
do you think it's better to call this parameter `handler_name`?

Comment (shubham-shay-te) [reply website/docs/cache.md:83]: yeah, that will be a good idea, I'll make the change

Comment (shay-te) [reply website/docs/cache.md:83]: here it still says `handler`

**Review by shay-te — CHANGES_REQUESTED** [review]

**Review by shay-te — APPROVED** [review]
